fix(eslint-plugin): respect custom pageExtensions in no-html-link-for-pages rule

Rezakarimzadeh98 opened the pull request:
## What?

This PR fixes the `@next/next/no-html-link-for-pages` ESLint rule to properly recognize pages when using custom `pageExtensions` in Next.js configuration files.

## Why?

Currently, the ESLint rule is hardcoded to only recognize `.js`, `.jsx`, `.ts`, and `.tsx` files as pages. When users configure custom page extensions (e.g., `.page.tsx`, `.mdx`, `.vue`) in their `next.config.js`, the rule fails to detect these pages and produces false positives, incorrectly flagging valid internal links as errors.

This issue has been reported by multiple users:
- Issue #53473 (15+ 👍 reactions indicating high demand)
- Existing PR #80035 has been open for 6 months but lacks TypeScript config support

## How?

### Key Changes

**1. Dynamic Extension Loading (`packages/eslint-plugin-next/src/utils/url.ts`)**

Added `getPageExtensions()` function that:
- Reads `pageExtensions` from `next.config.js`, `next.config.mjs`, or `next.config.ts`
- Uses closure-based memoization for performance (reads config only once)
- Supports TypeScript configs via `tsx` or `ts-node` loaders
- Falls back to default extensions `['tsx', 'ts', 'jsx', 'js']` if config is not found

```typescript
const getPageExtensions = (() => {
  let cached: string[] | null = null
  return (): string[] => {
    if (cached) return cached
    // ... reads from next.config.js/mjs/ts
    cached = extensions
    return cached
  }
})()

Commit:
fix(eslint-plugin): respect custom pageExtensions in no-html-link-for-pages rule

- Add support for next.config.js, next.config.mjs, and next.config.ts
- Dynamically load pageExtensions from config with fallback to defaults
- Fix isDirectory() bug (was incorrectly called with dirPath argument)
- Remove hardcoded regex patterns for page extensions
- Use memoization for performance optimization

Fixes #53473

Commit:
test: add comprehensive tests for getPageExtensions

- Add tests for custom extensions (.page.tsx, .mdx)
- Add memoization verification test
- Add changeset for version bump

Commit:
fix: remove unnecessary type assertions for optional dependencies

ijjk commented:
Allow CI Workflow Run

- [ ] approve CI run for commit: 2b973b9b3598aeba670b370e917d0a1ddc7dfa9a

Note: this should only be enabled once the PR is ready to go and can only be enabled by a maintainer

ijjk commented:
Allow CI Workflow Run

- [ ] approve CI run for commit: 5fc081449b99cefb43278b75879154c6a891c27e

Note: this should only be enabled once the PR is ready to go and can only be enabled by a maintainer

Commit:
chore: remove unrelated next-env.d.ts file

Commit:
fix: update tests to mock fs.existsSync correctly

Commit:
feat: add support for custom pageExtensions in no-html-link-for-pages rule

- Added getPageExtensions() function to read pageExtensions from next.config
- Added reset functions for caches to support testing
- Added comprehensive test for custom pageExtensions
- Fixed TypeScript type error in url.test.ts
- Supports next.config.js, next.config.mjs, and next.config.ts

Rezakarimzadeh98 committed:
Merge branch 'canary' into fix/eslint-pageextensions-support

Rezakarimzadeh98 committed:
Merge branch 'canary' into fix/eslint-pageextensions-support

Rezakarimzadeh98 committed:
Merge branch 'canary' into fix/eslint-pageextensions-support

Rezakarimzadeh98 committed:
fix: resolve critical issues in pageExtensions implementation

- Fix cache pollution by converting closure-based cache to Map-based cache keyed by cwd
- Fix duplicate routes bug by adding else statement for index files
- Fix incorrect recursive call in parseUrlForAppDir (was calling parseUrlForPages)

Addresses feedback from Copilot AI and Graphite bot reviews on PR #86408

Rezakarimzadeh98 committed:
test: add missing fs.existsSync mocks to tests

Addresses Vercel bot feedback - tests for invalid pageExtensions
and empty array now properly mock fs.existsSync to ensure they
test the intended behavior rather than file not found scenario.

Rezakarimzadeh98 committed:
Merge branch 'canary' into fix/eslint-pageextensions-support

Rezakarimzadeh98 committed:
Merge branch 'canary' into fix/eslint-pageextensions-support

Rezakarimzadeh98 committed:
Merge branch 'canary' into fix/eslint-pageextensions-support

Rezakarimzadeh98 committed:
Merge remote-tracking branch 'upstream/canary' into fix/eslint-pageextensions-support

Rezakarimzadeh98 committed:
chore(eslint-plugin): remove unrelated reproduction files

Rezakarimzadeh98 committed:
fix(eslint-plugin): resolve pageExtensions per root dir

Rezakarimzadeh98 committed:
fix(eslint-plugin): avoid re-registering ts loader

Rezakarimzadeh98 commented:
## ✅ Update: All Issues Resolved

Hi team! 👋

I've addressed all automated review feedback:

### Fixed Issues:
- ✅ **Cache pollution** - Using Map<rootDir, extensions> instead of single cache
- ✅ **Recursive bug** - parseUrlForAppDir now correctly calls itself
- ✅ **Index files** - Already using if-else to prevent duplicates
- ✅ **Test fixtures** - Created missing test directory structure
- ✅ **Test mocks** - Fixed fs.existsSync mocks in all tests

### Implementation Summary:
- Dynamic pageExtensions loading from next.config.js/mjs/ts
- Per-project caching (no cross-project pollution)
- TypeScript config support (tsx/ts-node)
- Full backward compatibility
- Comprehensive test coverage

@ijjk Ready for CI approval! 🚀

Rezakarimzadeh98 commented:
@ijjk Ready for CI approval! 🚀

All review feedback has been addressed and the code is ready for testing. Looking forward to getting this merged!
